### PR TITLE
story/RWA-449 - Improve SDK Error handling

### PR DESCRIPTION
**Pull request by HG-rc**
Ticket: https://rotacloud.atlassian.net/browse/RWA-449

Makes the error created in the SDK more developer friendly (& similar to the old LegacyApiError in the app).
Also reduces (but does not yet remove entirely) the need for boilerplate code when handling errors on the fetch response.

Having discussed with @liammoore4 we may be able to remove the promise handlers by moving to async/await & removing models in favour of interfaces at a later date.

### CODE REVIEW

File: src/services/service.ts [hunk 1]
```diff
@@ -153,6 +175,7 @@ export abstract class Service<ApiResponse = any> {
     let pageRemaining = true;
     while (pageRemaining) {
       const res = await this.fetch<T[]>(pageRequestObject, options);
+      console.log('this shouldnt happen', currentPageUrl);
```
Comment (liammoore4) [inline src/services/service.ts:178]: this shouldnt be here

Comment (HG-rc) [reply src/services/service.ts]: removed

**Comment by liammoore4** [comment]
Also forgot to mention but is the error type exported out of the SDK now?

File: src/models/SDKError.model.ts [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+export class SDKError extends Error {
+  readonly code?: number;
+  readonly data?: unknown;
+  readonly originalError: Error;
+
+  constructor(params: SDKErrorParams) {
+    super(params.message);
+
+    this.name = 'SDKError';
+    Error.captureStackTrace?.(this, SDKError);
```
Comment (richwatts0) [inline src/models/SDKError.model.ts:15]: Do we need the optional chaining operator here? I also wouldn't call the arguments for this method `params`

File: src/models/SDKError.model.ts [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+}
+export class SDKError extends Error {
+  readonly code?: number;
+  readonly data?: unknown;
+  readonly originalError: Error;
+
+  constructor(params: SDKErrorParams) {
+    super(params.message);
+
+    this.name = 'SDKError';
```
[the remark this answers is not in the document]
Comment (HG-rc) [reply src/models/SDKError.model.ts]: updated

File: src/services/service.ts [hunk 4]
```diff
@@ -59,7 +60,28 @@ type ParameterPrimitive = string | boolean | number | null;
...
+  protected client: AxiosInstance = this.initialiseAxios();
+
+  private initialiseAxios(): AxiosInstance {
+    const client = axios.create();
+    client.interceptors.response.use(
+      (response) => response,
+      (error) => {
+        let newError = error;
+        if (error.isAxiosError) {
+          const axiosResponse = error.response;
```
[the remark this answers is not in the document]
Comment (HG-rc) [reply src/services/service.ts]: Have changed this a bit to check for the request if the response isn't available - was curious if we should maybe include a similar interceptor on the request - seems to be allowed but any errors I could trigger would always be picked up in the response.

File: src/models/SDKError.model.ts (hunk 2 above)
Comment (HG-rc) [reply src/models/SDKError.model.ts]: Removed the chaining operator, renamed the args & matching interface

Comment (liammoore4) [reply src/models/SDKError.model.ts]: Think we need the optional chain here - this is a V8 specific method so may not be callable in other engines
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error#static_methods

**Review by liammoore4 — APPROVED** [review]

**Review by richwatts0 — APPROVED** [review]

**Pull request merged**